File: src/mem/bus.zig
//! The memory bus: turns a 32-bit address into either RAM bytes or a peripheral access.
//!
//! Phase 1 is a *flat* model. Two RAM regions (IRAM and DRAM) sized and placed by the SoC
//! profile, plus one MMIO decode for UART0. No cache, no MMU, no flash, no address
//! aliasing between the instruction and data buses.
//!
//! Endianness: the ESP32's Xtensa core is little-endian, so a 32-bit word at address A is
//! stored as bytes A=LSB .. A+3=MSB. We use `std.mem.readInt`/`writeInt` with an explicit
//! `.little` so the emulator behaves identically on a big-endian host.

const std = @import("std");
const soc = @import("../soc/esp32.zig");
const Uart = @import("../periph/uart.zig").Uart;

/// Everything that can go wrong on a bus access.
///
/// `error{...}` declares an *error set*. A function returning `BusError!u32` returns
/// either a u32 or one of these errors — Zig's error union. There are no exceptions and
/// no null-on-failure; the caller must handle or propagate with `try`.
pub const BusError = error{
    /// No RAM region and no known peripheral covers this address.
    UnmappedAddress,
    /// A 16- or 32-bit access whose address is not a multiple of the access size.
    /// Real Xtensa raises a LoadStoreAlignmentCause exception here; Phase 1 has no
    /// exception machinery, so we surface it as a bus error and halt.
    UnalignedAccess,
    /// The access started inside a region but ran off its end.
    RegionOverrun,
    /// A peripheral register block exists but this offset has no model.
    UnimplementedPeripheral,
};

pub const Bus = struct {
    /// Backing store for the two RAM regions. These are *slices* — a pointer and a
    /// length — pointing at memory owned by the allocator passed to `init`.
    iram: []u8,
    dram: []u8,

    /// The one peripheral Phase 1 models. A pointer, not a value, so writes made through
    /// the bus are visible to whoever owns the UART.
    uart: *Uart,

    /// Allocate the RAM regions. The caller must call `deinit` with the same allocator.
    pub fn init(allocator: std.mem.Allocator, uart: *Uart) std.mem.Allocator.Error!Bus {
        const iram = try allocator.alloc(u8, soc.iram_size);
        errdefer allocator.free(iram);
        const dram = try allocator.alloc(u8, soc.dram_size);

        // Real SRAM powers up with undefined contents, but an emulator that starts from
        // garbage is not reproducible. Zeroing makes every run identical, which is the
        // whole point of this project.
        @memset(iram, 0);
        @memset(dram, 0);

        return .{ .iram = iram, .dram = dram, .uart = uart };
    }

    pub fn deinit(self: *Bus, allocator: std.mem.Allocator) void {
        allocator.free(self.iram);
        allocator.free(self.dram);
        self.* = undefined;
    }

    // -----------------------------------------------------------------------------------
    // Region resolution
    // -----------------------------------------------------------------------------------

    /// Return the `len` bytes of RAM starting at `addr`, or null if `addr` is not RAM.
    /// Returns `RegionOverrun` if the access starts in RAM but crosses the region's end.
    fn ramSlice(self: *Bus, addr: u32, len: u32) BusError!?[]u8 {
        if (soc.isIram(addr)) return try sliceWithin(self.iram, addr - soc.iram_base, len);
        if (soc.isDram(addr)) return try sliceWithin(self.dram, addr - soc.dram_base, len);
        return null;
    }

    fn sliceWithin(region: []u8, offset: u32, len: u32) BusError![]u8 {
        if (@as(u64, offset) + len > region.len) return error.RegionOverrun;
        // `region[a..][0..n]` slices from `a` to the end, then takes the first `n` bytes.
        // It is the idiomatic way to get a fixed-length view for std.mem.readInt.
        return region[offset..][0..len];
    }

    // -----------------------------------------------------------------------------------
    // Loads
    // -----------------------------------------------------------------------------------

    pub fn read8(self: *Bus, addr: u32) BusError!u8 {
        if (try self.ramSlice(addr, 1)) |bytes| return bytes[0];
        // Peripheral registers on the ESP32 are word-oriented; a byte read of a register
        // returns the corresponding byte of the 32-bit value.
        const word = try self.readPeripheral(addr & ~@as(u32, 3));
        const shift: u5 = @intCast((addr & 3) * 8);
        return @truncate(word >> shift);
    }

    pub fn read16(self: *Bus, addr: u32) BusError!u16 {
        if (addr % 2 != 0) return error.UnalignedAccess;
        if (try self.ramSlice(addr, 2)) |bytes| {
            return std.mem.readInt(u16, bytes[0..2], .little);
        }
        const word = try self.readPeripheral(addr & ~@as(u32, 3));
        const shift: u5 = @intCast((addr & 2) * 8);
        return @truncate(word >> shift);
    }

    pub fn read32(self: *Bus, addr: u32) BusError!u32 {
        if (addr % 4 != 0) return error.UnalignedAccess;
        if (try self.ramSlice(addr, 4)) |bytes| {
            return std.mem.readInt(u32, bytes[0..4], .little);
        }
        return self.readPeripheral(addr);
    }

    // -----------------------------------------------------------------------------------
    // Stores
    // -----------------------------------------------------------------------------------

    pub fn write8(self: *Bus, addr: u32, value: u8) BusError!void {
        if (try self.ramSlice(addr, 1)) |bytes| {
            bytes[0] = value;
            return;
        }
        // A byte store to a peripheral is modelled as a word store of the byte value.
        // Only UART_FIFO_REG cares, and it only looks at the low 8 bits.
        return self.writePeripheral(addr & ~@as(u32, 3), value);
    }

    pub fn write16(self: *Bus, addr: u32, value: u16) BusError!void {
        if (addr % 2 != 0) return error.UnalignedAccess;
        if (try self.ramSlice(addr, 2)) |bytes| {
            std.mem.writeInt(u16, bytes[0..2], value, .little);
            return;
        }
        return self.writePeripheral(addr & ~@as(u32, 3), value);
    }

    pub fn write32(self: *Bus, addr: u32, value: u32) BusError!void {
        if (addr % 4 != 0) return error.UnalignedAccess;
        if (try self.ramSlice(addr, 4)) |bytes| {
            std.mem.writeInt(u32, bytes[0..4], value, .little);
            return;
        }
        return self.writePeripheral(addr, value);
    }

    // -----------------------------------------------------------------------------------
    // MMIO dispatch
    // -----------------------------------------------------------------------------------

    fn readPeripheral(self: *Bus, addr: u32) BusError!u32 {
        if (soc.isUart0(addr)) return self.uart.readReg(addr - soc.uart0_base);
        if (soc.isPeripheral(addr)) return error.UnimplementedPeripheral;
        return error.UnmappedAddress;
    }

    fn writePeripheral(self: *Bus, addr: u32, value: u32) BusError!void {
        if (soc.isUart0(addr)) return self.uart.writeReg(addr - soc.uart0_base, value);
        if (soc.isPeripheral(addr)) return error.UnimplementedPeripheral;
        return error.UnmappedAddress;
    }

    // -----------------------------------------------------------------------------------
    // Bulk access, used by the ELF loader
    // -----------------------------------------------------------------------------------

    /// Copy `bytes` into RAM starting at `addr`. Peripherals are not loadable.
    pub fn writeBlock(self: *Bus, addr: u32, bytes: []const u8) BusError!void {
        const dest = (try self.ramSlice(addr, @intCast(bytes.len))) orelse
            return error.UnmappedAddress;
        @memcpy(dest, bytes);
    }

    /// Fill `len` bytes of RAM at `addr` with zero. Used for the .bss part of a segment,
    /// where the ELF says memsz > filesz.
    pub fn zeroBlock(self: *Bus, addr: u32, len: u32) BusError!void {
        const dest = (try self.ramSlice(addr, len)) orelse return error.UnmappedAddress;
        @memset(dest, 0);
    }
};

// ---------------------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------------------

const testing = std.testing;

/// Small helper so each test does not repeat the setup. Zig has no test fixtures; a
/// struct plus explicit `start`/`deinit` calls is the usual pattern.
///
/// Note that `start` is separate from a constructor: `bus` holds a pointer to `uart`, and
/// a struct returned *by value* from a constructor would move in memory, leaving that
/// pointer dangling. Taking `&self.uart` only after the Harness has its final address is
/// the safe order. Zig will not catch this mistake for you.
const Harness = struct {
    uart: Uart = .{},
    bus: Bus = undefined,

    fn start(self: *Harness) !void {
        self.bus = try Bus.init(testing.allocator, &self.uart);
    }

    fn deinit(self: *Harness) void {
        self.bus.deinit(testing.allocator);
    }
};

test "word round-trip in dram" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    try h.bus.write32(soc.dram_base + 16, 0xDEAD_BEEF);
    try testing.expectEqual(@as(u32, 0xDEAD_BEEF), try h.bus.read32(soc.dram_base + 16));
}

test "little-endian byte order" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    try h.bus.write32(soc.dram_base, 0x1122_3344);
    // Least significant byte first.
    try testing.expectEqual(@as(u8, 0x44), try h.bus.read8(soc.dram_base + 0));
    try testing.expectEqual(@as(u8, 0x33), try h.bus.read8(soc.dram_base + 1));
    try testing.expectEqual(@as(u8, 0x22), try h.bus.read8(soc.dram_base + 2));
    try testing.expectEqual(@as(u8, 0x11), try h.bus.read8(soc.dram_base + 3));
    try testing.expectEqual(@as(u16, 0x3344), try h.bus.read16(soc.dram_base + 0));
    try testing.expectEqual(@as(u16, 0x1122), try h.bus.read16(soc.dram_base + 2));
}

test "iram and dram are independent regions" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    try h.bus.write32(soc.iram_base, 0xAAAA_AAAA);
    try h.bus.write32(soc.dram_base, 0xBBBB_BBBB);
    try testing.expectEqual(@as(u32, 0xAAAA_AAAA), try h.bus.read32(soc.iram_base));
    try testing.expectEqual(@as(u32, 0xBBBB_BBBB), try h.bus.read32(soc.dram_base));
}

test "alignment is enforced for halfword and word access" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    try testing.expectError(error.UnalignedAccess, h.bus.read32(soc.dram_base + 1));
    try testing.expectError(error.UnalignedAccess, h.bus.read32(soc.dram_base + 2));
    try testing.expectError(error.UnalignedAccess, h.bus.read16(soc.dram_base + 1));
    try testing.expectError(error.UnalignedAccess, h.bus.write32(soc.dram_base + 3, 0));
    // Byte access is never misaligned.
    try h.bus.write8(soc.dram_base + 1, 0x5A);
    try testing.expectEqual(@as(u8, 0x5A), try h.bus.read8(soc.dram_base + 1));
}

test "unmapped addresses are an error, not silent zeroes" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    try testing.expectError(error.UnmappedAddress, h.bus.read32(0x0000_0000));
    try testing.expectError(error.UnmappedAddress, h.bus.read32(0x5000_0000));
    try testing.expectError(error.UnmappedAddress, h.bus.write32(soc.iram_base - 4, 1));
}

test "an access that runs off the end of a region is caught" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    const last_word = soc.dram_base + soc.dram_size - 4;
    try h.bus.write32(last_word, 0x1234_5678);
    try testing.expectEqual(@as(u32, 0x1234_5678), try h.bus.read32(last_word));

    // One word past the end is not DRAM at all.
    try testing.expectError(error.UnmappedAddress, h.bus.read32(last_word + 4));

    // A bulk write that starts inside DRAM but overruns its end. (Single loads and
    // stores cannot trigger this: alignment is enforced and the region size is a
    // multiple of 4, so an aligned access either fits entirely or starts outside.)
    var scratch: [8]u8 = undefined;
    try testing.expectError(error.RegionOverrun, h.bus.writeBlock(last_word, &scratch));
}

test "unimplemented peripherals are reported, not ignored" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    // GPIO base on the ESP32; no model in Phase 1.
    try testing.expectError(error.UnimplementedPeripheral, h.bus.write32(0x3FF4_4004, 1));
}

test "a word store to the uart fifo emits a byte" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    try h.bus.write32(soc.uart0_base + soc.uart_reg.fifo, 'Z');
    try testing.expectEqualStrings("Z", h.uart.transmitted());
}

test "writeBlock and zeroBlock cover a segment load" {
    var h = Harness{};
    try h.start();
    defer h.deinit();

    try h.bus.writeBlock(soc.iram_base, &[_]u8{ 1, 2, 3, 4 });
    try testing.expectEqual(@as(u32, 0x0403_0201), try h.bus.read32(soc.iram_base));

    try h.bus.zeroBlock(soc.iram_base, 4);
    try testing.expectEqual(@as(u32, 0), try h.bus.read32(soc.iram_base));

    // A segment that does not fit in any region must fail loudly.
    try testing.expectError(error.UnmappedAddress, h.bus.writeBlock(0x1000, &[_]u8{0}));
}

File: crates/simulatr-core/src/mem/bus.rs
//! The memory bus: turns a 32-bit address into RAM bytes or a peripheral access.
//!
//! Phase 1 is a flat model — two RAM regions plus one MMIO decode, no cache, no MMU, no
//! flash, no aliasing between the instruction and data buses. All address decoding is
//! delegated to the SoC profile; this module only routes.
//!
//! The ESP32's Xtensa core is little-endian, and every multi-byte access here goes
//! through `from_le_bytes`/`to_le_bytes` so the emulator behaves identically on a
//! big-endian host.

use crate::periph::uart::Uart;
use crate::soc::esp32::{self, Mapping};
use core::fmt;

/// Access width. The discriminants are the byte counts, which is also the alignment
/// requirement — Xtensa has no unaligned 16- or 32-bit access.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Width {
    Byte = 1,
    Half = 2,
    Word = 4,
}

impl Width {
    pub const fn bytes(self) -> u32 {
        self as u32
    }
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum BusError {
    /// No RAM region and no known peripheral covers this address.
    Unmapped { addr: u32 },
    /// A real peripheral register block with no model in this phase.
    UnimplementedPeripheral { addr: u32 },
    /// A 16- or 32-bit access whose address is not a multiple of the access size. Real
    /// Xtensa raises `LoadStoreAlignmentCause`; Phase 1 has no exception machinery, so
    /// this halts the run instead.
    Unaligned { addr: u32, width: Width },
    /// The access started inside a region but ran off its end. Single loads and stores
    /// cannot trigger this (alignment is enforced and region sizes are multiples of 4),
    /// so in practice it only comes from a bulk segment load.
    RegionOverrun { addr: u32, len: u32 },
}

impl fmt::Display for BusError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            BusError::Unmapped { addr } => write!(f, "unmapped address {addr:#010x}"),
            BusError::UnimplementedPeripheral { addr } => {
                write!(f, "unimplemented peripheral at {addr:#010x}")
            }
            BusError::Unaligned { addr, width } => {
                write!(f, "unaligned {}-byte access at {addr:#010x}", width.bytes())
            }
            BusError::RegionOverrun { addr, len } => {
                write!(f, "{len}-byte access at {addr:#010x} runs off the end of its region")
            }
        }
    }
}

impl std::error::Error for BusError {}

pub struct Bus {
    iram: Box<[u8]>,
    dram: Box<[u8]>,
    uart: Uart,
}

/// Deliberately not derived: the RAM regions are half a megabyte and dumping them helps
/// nobody.
impl fmt::Debug for Bus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Bus")
            .field(
                "iram",
                &format_args!("{} bytes @ {:#010x}", self.iram.len(), esp32::IRAM_BASE),
            )
            .field(
                "dram",
                &format_args!("{} bytes @ {:#010x}", self.dram.len(), esp32::DRAM_BASE),
            )
            .field("uart", &self.uart)
            .finish()
    }
}

impl Default for Bus {
    fn default() -> Self {
        Self::new()
    }
}

impl Bus {
    pub fn new() -> Self {
        // Real SRAM powers up with garbage, but a simulator that starts from garbage is
        // not reproducible, and determinism is the product.
        Self {
            iram: vec![0; esp32::IRAM_SIZE as usize].into_boxed_slice(),
            dram: vec![0; esp32::DRAM_SIZE as usize].into_boxed_slice(),
            uart: Uart::new(),
        }
    }

    pub fn uart(&self) -> &Uart {
        &self.uart
    }

    pub fn uart_mut(&mut self) -> &mut Uart {
        &mut self.uart
    }

    // --- loads ---------------------------------------------------------------------

    pub fn read8(&self, addr: u32) -> Result<u8, BusError> {
        Ok(self.load(addr, Width::Byte)? as u8)
    }

    pub fn read16(&self, addr: u32) -> Result<u16, BusError> {
        Ok(self.load(addr, Width::Half)? as u16)
    }

    pub fn read32(&self, addr: u32) -> Result<u32, BusError> {
        self.load(addr, Width::Word)
    }

    /// One implementation for all three widths; the result is zero-extended into a u32.
    fn load(&self, addr: u32, width: Width) -> Result<u32, BusError> {
        let size = width.bytes();
        if addr % size != 0 {
            return Err(BusError::Unaligned { addr, width });
        }

        match esp32::decode(addr) {
            Mapping::Iram { offset } => read_le(&self.iram, offset, size, addr),
            Mapping::Dram { offset } => read_le(&self.dram, offset, size, addr),
            // Peripheral registers are word-oriented; a sub-word read returns the
            // corresponding slice of the 32-bit register value.
            Mapping::Uart0 { offset } => {
                let word = self.uart.read_reg(offset & !3);
                Ok(sub_word(word, addr, size))
            }
            Mapping::UnmappedPeripheral => Err(BusError::UnimplementedPeripheral { addr }),
            Mapping::Unmapped => Err(BusError::Unmapped { addr }),
        }
    }

    // --- stores --------------------------------------------------------------------

    pub fn write8(&mut self, addr: u32, value: u8) -> Result<(), BusError> {
        self.store(addr, Width::Byte, u32::from(value))
    }

    pub fn write16(&mut self, addr: u32, value: u16) -> Result<(), BusError> {
        self.store(addr, Width::Half, u32::from(value))
    }

    pub fn write32(&mut self, addr: u32, value: u32) -> Result<(), BusError> {
        self.store(addr, Width::Word, value)
    }

    fn store(&mut self, addr: u32, width: Width, value: u32) -> Result<(), BusError> {
        let size = width.bytes();
        if addr % size != 0 {
            return Err(BusError::Unaligned { addr, width });
        }

        match esp32::decode(addr) {
            Mapping::Iram { offset } => write_le(&mut self.iram, offset, size, addr, value),
            Mapping::Dram { offset } => write_le(&mut self.dram, offset, size, addr, value),
            // A sub-word store to a peripheral is modelled as a word store of the value.
            // Only UART_FIFO_REG cares, and it only looks at the low 8 bits.
            Mapping::Uart0 { offset } => {
                self.uart.write_reg(offset & !3, value);
                Ok(())
            }
            Mapping::UnmappedPeripheral => Err(BusError::UnimplementedPeripheral { addr }),
            Mapping::Unmapped => Err(BusError::Unmapped { addr }),
        }
    }

    // --- bulk access, used by the ELF loader ----------------------------------------

    /// Copy `bytes` into RAM at `addr`. Peripherals are not loadable.
    pub fn write_block(&mut self, addr: u32, bytes: &[u8]) -> Result<(), BusError> {
        let len = bytes.len() as u32;
        let dest = self.ram_block_mut(addr, len)?;
        dest.copy_from_slice(bytes);
        Ok(())
    }

    /// Zero `len` bytes of RAM at `addr` — the `.bss` tail of a segment, where the ELF
    /// says `p_memsz > p_filesz`.
    pub fn zero_block(&mut self, addr: u32, len: u32) -> Result<(), BusError> {
        self.ram_block_mut(addr, len)?.fill(0);
        Ok(())
    }

    fn ram_block_mut(&mut self, addr: u32, len: u32) -> Result<&mut [u8], BusError> {
        let (region, offset) = match esp32::decode(addr) {
            Mapping::Iram { offset } => (&mut self.iram, offset),
            Mapping::Dram { offset } => (&mut self.dram, offset),
            _ => return Err(BusError::Unmapped { addr }),
        };
        let end = offset as usize + len as usize;
        region
            .get_mut(offset as usize..end)
            .ok_or(BusError::RegionOverrun { addr, len })
    }
}

fn read_le(mem: &[u8], offset: u32, size: u32, addr: u32) -> Result<u32, BusError> {
    let start = offset as usize;
    let bytes = mem
        .get(start..start + size as usize)
        .ok_or(BusError::RegionOverrun { addr, len: size })?;

    let mut word = [0u8; 4];
    word[..bytes.len()].copy_from_slice(bytes);
    Ok(u32::from_le_bytes(word))
}

fn write_le(
    mem: &mut [u8],
    offset: u32,
    size: u32,
    addr: u32,
    value: u32,
) -> Result<(), BusError> {
    let start = offset as usize;
    let bytes = mem
        .get_mut(start..start + size as usize)
        .ok_or(BusError::RegionOverrun { addr, len: size })?;
    bytes.copy_from_slice(&value.to_le_bytes()[..size as usize]);
    Ok(())
}

/// Extract the `size` bytes at `addr`'s offset within a 32-bit register value.
fn sub_word(word: u32, addr: u32, size: u32) -> u32 {
    if size == 4 {
        return word;
    }
    let shift = (addr & 3) * 8;
    let mask = (1u64 << (size * 8)) - 1;
    ((word >> shift) as u64 & mask) as u32
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::soc::esp32::{DRAM_BASE, DRAM_SIZE, IRAM_BASE, UART0_BASE, uart_reg};

    #[test]
    fn word_round_trip_in_dram() {
        let mut bus = Bus::new();
        bus.write32(DRAM_BASE + 16, 0xDEAD_BEEF).unwrap();
        assert_eq!(bus.read32(DRAM_BASE + 16).unwrap(), 0xDEAD_BEEF);
    }

    #[test]
    fn little_endian_byte_order() {
        let mut bus = Bus::new();
        bus.write32(DRAM_BASE, 0x1122_3344).unwrap();

        assert_eq!(bus.read8(DRAM_BASE).unwrap(), 0x44);
        assert_eq!(bus.read8(DRAM_BASE + 1).unwrap(), 0x33);
        assert_eq!(bus.read8(DRAM_BASE + 2).unwrap(), 0x22);
        assert_eq!(bus.read8(DRAM_BASE + 3).unwrap(), 0x11);
        assert_eq!(bus.read16(DRAM_BASE).unwrap(), 0x3344);
        assert_eq!(bus.read16(DRAM_BASE + 2).unwrap(), 0x1122);
    }

    #[test]
    fn iram_and_dram_are_independent() {
        let mut bus = Bus::new();
        bus.write32(IRAM_BASE, 0xAAAA_AAAA).unwrap();
        bus.write32(DRAM_BASE, 0xBBBB_BBBB).unwrap();
        assert_eq!(bus.read32(IRAM_BASE).unwrap(), 0xAAAA_AAAA);
        assert_eq!(bus.read32(DRAM_BASE).unwrap(), 0xBBBB_BBBB);
    }

    #[test]
    fn alignment_is_enforced_for_halfword_and_word_access() {
        let mut bus = Bus::new();
        assert_eq!(
            bus.read32(DRAM_BASE + 1),
            Err(BusError::Unaligned { addr: DRAM_BASE + 1, width: Width::Word })
        );
        assert_eq!(
            bus.read32(DRAM_BASE + 2),
            Err(BusError::Unaligned { addr: DRAM_BASE + 2, width: Width::Word })
        );
        assert_eq!(
            bus.read16(DRAM_BASE + 1),
            Err(BusError::Unaligned { addr: DRAM_BASE + 1, width: Width::Half })
        );
        assert!(bus.write32(DRAM_BASE + 3, 0).is_err());

        // Byte access is never misaligned.
        bus.write8(DRAM_BASE + 1, 0x5A).unwrap();
        assert_eq!(bus.read8(DRAM_BASE + 1).unwrap(), 0x5A);
    }

    #[test]
    fn unmapped_addresses_are_errors_not_silent_zeroes() {
        let mut bus = Bus::new();
        assert_eq!(bus.read32(0), Err(BusError::Unmapped { addr: 0 }));
        assert_eq!(bus.read32(0x5000_0000), Err(BusError::Unmapped { addr: 0x5000_0000 }));
        assert!(bus.write32(IRAM_BASE - 4, 1).is_err());
    }

    #[test]
    fn unimplemented_peripherals_are_reported_not_ignored() {
        let mut bus = Bus::new();
        // GPIO base on the ESP32; no model in Phase 1.
        assert_eq!(
            bus.write32(0x3FF4_4004, 1),
            Err(BusError::UnimplementedPeripheral { addr: 0x3FF4_4004 })
        );
    }

    #[test]
    fn a_word_store_to_the_uart_fifo_emits_a_byte() {
        let mut bus = Bus::new();
        bus.write32(UART0_BASE + uart_reg::FIFO, u32::from(b'Z')).unwrap();
        assert_eq!(bus.uart().pending(), b"Z");
    }

    #[test]
    fn bulk_access_covers_a_segment_load() {
        let mut bus = Bus::new();
        bus.write_block(IRAM_BASE, &[1, 2, 3, 4]).unwrap();
        assert_eq!(bus.read32(IRAM_BASE).unwrap(), 0x0403_0201);

        bus.zero_block(IRAM_BASE, 4).unwrap();
        assert_eq!(bus.read32(IRAM_BASE).unwrap(), 0);

        // A segment aimed at an address with no RAM behind it must fail loudly.
        assert!(bus.write_block(0x1000, &[0]).is_err());
    }

    #[test]
    fn a_bulk_write_that_overruns_its_region_is_caught() {
        let mut bus = Bus::new();
        let last_word = DRAM_BASE + DRAM_SIZE - 4;
        bus.write32(last_word, 0x1234_5678).unwrap();
        assert_eq!(bus.read32(last_word).unwrap(), 0x1234_5678);
        assert_eq!(bus.read32(last_word + 4), Err(BusError::Unmapped { addr: last_word + 4 }));

        assert_eq!(
            bus.write_block(last_word, &[0; 8]),
            Err(BusError::RegionOverrun { addr: last_word, len: 8 })
        );
    }
}
